Improved anti-gravity

ctzsnooze opened the pull request:
Summary:  A high-pass filtered throttle parameter is used as a control value to multiply iTerm during fast throttle movements.

Purpose: to better control nose up/down wobble with rapid throttle change, and for that code to be more independent of Rx intervals.

Advantages over current Anti-Gravity code:

- No threshold required

- Smoothly enhances I as rate of change of throttle increases

- Works well regardless of different Rx step intervals, because no threshold is needed, and because the output of the high-pass doesn't change much with different Rx intervals.  

- More effective at reducing nose wobble with rapid changes of throttle than the old code.

- Works well with the pilot's previous antigravity gain setting

What was wrong with the old code?  

The old iTerm boost was only was active while the current RC step change was above threshold.  For the duration of that step, iTerm was multiplied by AG gain.  All the rest of the time, AG gain was 1.  It was an all or none thing.  In reality, the pitching tendency was not an all or none thing.  To exceed threshold required very fast stick input.  The size of the up-step was dependent on RC interval.  To exceed the 350 threshold with FrSky 9ms was possible, but required nearly a full throttle change in less than 50ms, which is a super fast stick movement.  So the actual antigravity effect was a bit random, either all or none, mostly none.  With TBS protocols where steps were 4-5ms apart, it would be impossible to trigger it.  So the old code didn't work with those fast protocols.  And with slow protocols, e.g. 18ms SBus, it would have triggered much more frequently than intended.  In practice it just didn't work as well as it should; this approach is smoother, more 'analog', and works much better.

Note that this code is currently on throttle boost, since it uses the same throttle hpf value to multiply iTerm.  It won't work on targets that do not support throttle boost.

ledvinap commented on src/main/flight/pid.c:
```diff
@@ -771,14 +774,32 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+    // ThrottleHpf typically reaches max of 0.1 for fast throttle changes
+    // AG Gain of 5000 results in itermAccelerateNew of 5 at throttleHpf of 0.1
+    // if throttle movement is zero, throttleHpf is zero, I accumulates normally
+        float itermAccelerateNew = (1 + (fabsf(throttleHpf) * 0.01 * (pidProfile->itermAcceleratorGain - 1000)));
```
Is there any need for negative gain? And even in that case itermAcceleratorGain shall be changes to signed, not offset by 1000 ...

ctzsnooze replied:
The range for gain in CLI is 1000 to some much bigger number.  1000 translates to a gain of 1.  In order to retain the old gain values, the subtraction was necessary.  The output of that line should be 1 when the user sets the gain to 1.

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -121,6 +121,7 @@ static FAST_RAM_ZERO_INIT float motorMixRange;
 
 float FAST_RAM_ZERO_INIT motor[MAX_SUPPORTED_MOTORS];
 float motor_disarmed[MAX_SUPPORTED_MOTORS];
+static FAST_RAM_ZERO_INIT float throttlehpf;
```
Better `throttleHpf`.

ctzsnooze replied:
In mixer.c JoeLucid named the internal throttle highness `throttlehpf`.  I didn't change any of that.

When passing to pid.c I used ThrottleHpf as the 'carrier', and inside pid.c I used throttleHpf.

I don't know about it being a good idea just changing the capitalisation... but that's what I did  :-)

Happy to modify as appropriate.

mikeller replied:
Naming convention for Betaflight is to use camelCase for variables, so `throttleHpf` fits with the convention, yes please.

ctzsnooze replied:
Fixed.  Thanks for the explanation.

mikeller commented on src/main/flight/pid.c:
```diff
@@ -771,14 +774,32 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+        dynCi = MIN((1.0f - motorMixRange) * ITermWindupPointInv, 1.0f) * dT * itermAccelerateNew;
+           DEBUG_SET(DEBUG_ITERM_RELAX, 0, lrintf(throttleHpf * 1000));
+           DEBUG_SET(DEBUG_ITERM_RELAX, 1, lrintf(itermAccelerateNew * 1000));
+    } else {
```
This can be made a bit more elegant:

ctzsnooze replied:
yes that's nicer

DieHertz commented on src/main/flight/mixer.c:
```diff
@@ -331,6 +332,11 @@ float getMotorMixRange(void)
 
+float getThrottleHpf(void)
+{
+return throttlehpf;
```
indentation missing

ctzsnooze replied:
sorry!  fixed

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -771,14 +774,32 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+    // ThrottleHpf typically reaches max of 0.1 for fast throttle changes
+    // AG Gain of 5000 results in itermAccelerateNew of 5 at throttleHpf of 0.1
+    // if throttle movement is zero, throttleHpf is zero, I accumulates normally
+        float itermAccelerateNew = (1 + (fabsf(throttleHpf) * 0.01 * (pidProfile->itermAcceleratorGain - 1000)));
```
There's a double literal `0.01` sneaking into BF code :)

ctzsnooze replied:
oh so it should be 0.01f ?  whoops

DieHertz commented on src/main/flight/pid.h:
```diff
@@ -126,6 +126,7 @@ typedef struct pidProfile_s {
     uint8_t throttle_boost;                 // how much should throttle be boosted during transient changes 0-100, 100 adds 10x hpf filtered throttle
     uint8_t throttle_boost_cutoff;          // Which cutoff frequency to use for throttle boost. higher cutoffs keep the boost on for shorter. Specified in hz.
+    uint8_t anti_gravity_new;               // Option to use new smooth anti-gravity that requires throttle high pass from throttle boost.
     uint8_t  iterm_rotation;                // rotates iterm to translate world errors to local coordinate system
```
While you're at it, could you fix alignment of two following fields? They seem to be separated with 2 spaces or one tab rather than 1 space.

ctzsnooze replied:
yes I will

DieHertz commented on src/main/interface/settings.c:
```diff
@@ -817,6 +817,7 @@ const clivalue_t valueTable[] = {
 #if defined(USE_THROTTLE_BOOST)
     { "throttle_boost",             VAR_UINT8 | PROFILE_VALUE,  .config.minmax = { 0, 100 }, PG_PID_PROFILE, offsetof(pidProfile_t, throttle_boost) },
     { "throttle_boost_cutoff",      VAR_UINT8 | PROFILE_VALUE,  .config.minmax = { 5, 50 }, PG_PID_PROFILE, offsetof(pidProfile_t, throttle_boost_cutoff) },
+    { "anti_gravity_new",           VAR_UINT8  | PROFILE_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_OFF_ON }, PG_PID_PROFILE, offsetof(pidProfile_t, anti_gravity_new) },
```
Two spaces after `VAR_UINT8`, should be 1

ctzsnooze replied:
thanks, fixed that and also spaces above for relax

mikeller commented on src/main/flight/pid.c:
```diff
@@ -771,14 +774,33 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+    {
+        // normal iTerm
+        dynCi = MIN((1.0f - motorMixRange) * ITermWindupPointInv, 1.0f) * dT * itermAccelerator;
+#if defined(USE_THROTTLE_BOOST)
```
You don't need this `#if defined...` any more - that's what's pretty about my suggestion. :-)

ctzsnooze replied:
So like this:
`    } else 
#endif
    {
        // normal iTerm
        dynCi = MIN((1.0f - motorMixRange) * ITermWindupPointInv, 1.0f) * dT * itermAccelerator;
    }
`

mikeller replied:
Yes.

ctzsnooze replied:
very clever  :-)  thank you!

ctzsnooze commented:
please review [97232c3](https://github.com/betaflight/betaflight/pull/6220/commits/97232c318c6b1c140ceba6f9af9f6379aa021443)
I hope to remove the dependency on throttle boost value being above zero...

etracer65 commented:
It also might be a good idea to position this a "new experimental" anti-gravity that can be used instead of the classic method - similar to how rc smoothing is an experimental replacement for interpolation.  Since nearly everybody uses anti-gravity we need to tread lightly and make a path forward.  Even if the code is 100% better and the right way to go making a wholesale flight dynamics change is tough as some people are sure to like the "way it was".  Maybe position it by having a `anti_gravity_type = CLASSIC | NEW` (or maybe a better name for "NEW") that controls which code gets used.  We can then decide when the type should change as a default after testing and at some point later retire the "CLASSIC" code.

ctzsnooze commented:
I've done what I can to separate this code from throttle boost in 97232c3.  I don't have the programming skills to separate it any better - please feel free to show me how.  

The old AG code is still in place, if the user turns off the new code (set anti_gravity_new = OFF), the old code will run instead (at least, that's the intention of the code).  The old threshold is still there, and the gain values are essentially interchangeable.  A-B testing is quite easy, and we have some time to test it.  To date all testing has been strongly positive.  I anticipate the old code should be removed in due course.

etracer65 commented:
Another aspect with mixing with throttle boost that's not desirable is that they share the same LPF and thus the same cutoff.  The user can change the cutoff for throttle boost which will (unknowingly to them) also change the behavior of the new anti-gravity.  As much as I hate to say it, I think anti-gravity should have it's own filter on throttle separate from throttle boost.  This will make separating the code simpler as well.

etracer65 commented on src/main/flight/pid.c:
```diff
@@ -771,14 +774,31 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+        float itermAccelerateNew = (1 + (fabsf(throttleHpf) * 0.01f * (pidProfile->itermAcceleratorGain - 1000)));
+        dynCi = MIN((1.0f - motorMixRange) * ITermWindupPointInv, 1.0f) * dT * itermAccelerateNew;
+           DEBUG_SET(DEBUG_ITERM_RELAX, 0, lrintf(throttleHpf * 1000));
+           DEBUG_SET(DEBUG_ITERM_RELAX, 1, lrintf(itermAccelerateNew * 1000));
```
Doesn't seem like this debugging should be under `DEBUG_ITERM_RELAX`.  Also indention is wrong.

etracer65 commented on src/main/flight/pid.c:
```diff
@@ -771,14 +774,31 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+    // ThrottleHpf typically reaches max of 0.1 for fast throttle changes
+    // AG Gain of 5000 results in itermAccelerateNew of 5 at throttleHpf of 0.1
+    // if throttle movement is zero, throttleHpf is zero, I accumulates normally
+        float itermAccelerateNew = (1 + (fabsf(throttleHpf) * 0.01f * (pidProfile->itermAcceleratorGain - 1000)));
```
Should be `const`

etracer65 commented on src/main/flight/pid.h:
```diff
@@ -126,8 +126,9 @@ typedef struct pidProfile_s {
     uint8_t throttle_boost_cutoff;          // Which cutoff frequency to use for throttle boost. higher cutoffs keep the boost on for shorter. Specified in hz.
-    uint8_t  iterm_rotation;                // rotates iterm to translate world errors to local coordinate system
-    uint8_t  smart_feedforward;             // takes only the larger of P and the D weight feed forward term if they have the same sign.
+    uint8_t anti_gravity_new;               // Option to use new smooth anti-gravity that requires throttle high pass from throttle boost.
```
By inserting this value in the middle and changing the structure you need to update the version on the PG in pid.c

etracer65 replied:
Or simply move the new parameter to the end of the structure (append rather than insert) and you won't need to update the PG version.

simonbethke commented:
Hey, nice to finally See this as a PR. I really would like to contribute a Version that doesn't even require a gain value. However, I don't find time for implementing and if I could find time for that there would be none for testing.
The approach I'd like to contribute is Based on the assumption, that the PID P-term setting already abstracts the weight Distribution around the COG on that axis. Using P-term setting instead of gain on that axis should work.

ctzsnooze commented:
eTracer's major tidy up is in place now - thanks so much for your help!

mikeller commented on src/main/fc/fc_core.c:
```diff
@@ -848,6 +848,8 @@ bool processRx(timeUs_t currentTimeUs)
     pidSetAcroTrainerState(IS_RC_MODE_ACTIVE(BOXACROTRAINER) && sensors(SENSOR_ACC));
 #endif // USE_ACRO_TRAINER
 
+    pidSetAntiGravityState(IS_RC_MODE_ACTIVE(BOXANTIGRAVITY) || feature(FEATURE_ANTI_GRAVITY));
```
Very minor: More efficient for default settings when `feature(FEATURE_ANTI_GRAVITY)` is checked first.

etracer65 replied:
I'll change this around.  I simply moved this check from fc_rc.c so it was previously this ordering.

mikeller commented on src/main/flight/pid.c:
```diff
@@ -180,9 +190,9 @@ void pidSetItermAccelerator(float newItermAccelerator)
 }
 
-float pidItermAccelerator(void)
+bool pidOSDAntiGravityActive(void)
```
camelCase is normally done by only capitalising the first letter of abbreviations, i.e. `pidOsdAntiGravityActive()`.

mikeller commented on src/main/flight/pid.h:
```diff
@@ -184,3 +184,7 @@ void pidAcroTrainerInit(void);
 void pidSetAcroTrainerState(bool newState);
 void pidInitSetpointDerivativeLpf(uint16_t filterCutoff, uint8_t debugAxis, uint8_t filterType);
 void pidUpdateSetpointDerivativeLpf(uint16_t filterCutoff);
+void pidUpdateAGThrottleFilter(float throttle);
```
`pidUpdateAntiGravityThrottleFilter()` would be more consistent.

etracer65 replied:
I originally had that but thought the function name was getting pretty long. ;)  I'll fix.

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -778,7 +812,19 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+        // antiGravityThrottleHpf typically reaches max of 0.1 for fast throttle changes
+        // AG Gain of 5000 results in itermAccelerateFactor of 5 at antiGravityThrottleHpf of 0.1
+        // if throttle movement is zero, antiGravityThrottleHpf is zero, I accumulates normally
+        itermAccelerator = (1 + (fabsf(antiGravityThrottleHpf) * 0.01f * (itermAcceleratorGain - 1000)));
```
The outermost parentheses are surely not needed. Parens around `fabsf() * .. * ..)` also don't make it any more clear IMO.

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -1021,3 +1067,16 @@ void pidSetAcroTrainerState(bool newState)
+
+void pidSetAntiGravityState(bool newState)
+{
+    if (newState != antiGravityEnabled) { // reset the accelerator on state changes
```
Please add code comments on a separate line, not inline

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -1021,3 +1067,16 @@ void pidSetAcroTrainerState(bool newState)
+bool pidAntiGravityEnabled(void)
+{
+    return antiGravityEnabled;
+}
```
No newline at the end of file

ctzsnooze replied:
fixed, put one at the end of mixer.c also

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -144,6 +153,7 @@ void resetPidProfile(pidProfile_t *pidProfile)
         .itermLimit = 150,
         .throttle_boost = 5,
         .throttle_boost_cutoff = 15,
+        .anti_gravity_new = true,
```
Since `anti_gravity_new` is not appended at the end of PG, let's synchronize the order of initializers as well, moving this line to the end.

theSkonk commented:
have flown a few versions of this and its nice to actually have AG that works and that i understand the implementation of the old code is basically ineffective for most people

jirif commented on src/main/flight/pid.h:
```diff
@@ -138,6 +138,7 @@ typedef struct pidProfile_s {
     uint8_t abs_control_gain;               // How strongly should the absolute accumulated error be corrected for
     uint8_t abs_control_limit;              // Limit to the correction
     uint8_t abs_control_error_limit;        // Limit to the accumulated error
+    uint8_t anti_gravity_new;               // Option to use new smooth anti-gravity that requires throttle high pass from throttle boost.
```
Please update the comment, now it have its own lpf.

etracer65 commented:
I see that we're now getting F3 target flash overflows so I presume this needs to be rebased.  Not sure what changed to cause the substantial overflows.  Or maybe this branch was made before some of the flash usage reductions were implemented.

mikeller commented:
@etracer65: Since we are targeting 4.0 with this, I think we can leave the target overflow for now, and look into how much slashing needs to be done when 3.4 is out and we are getting ready to merge the pull requests slated for 4.0.

etracer65 commented:
@mikeller No problem.  I'm still going to look into it a little when I get a chance. It seems like the targets suddenly grew when I was only making cosmetic and small changes - but I'm not completely sure about this. Haven't had time to get back to it yet.

mikeller commented:
@etracer65: Is that observation based on the outcome of travis builds? If so you are probably not aware that travis builds (with the new integration) are run on a speculative merge of the pull request branch into master at the time that the build was started:

> It is running a build against the merge commit, after merging #6220 Improved anti-gravity.
> Any changes that have been made to the master branch before the build ran are also included.

etracer65 commented:
I’m not 100% sure but I thought I saw the large overflow locally as well. Need to verify though.

McGiverGim commented on src/main/interface/settings.c:
```diff
@@ -817,6 +817,7 @@ const clivalue_t valueTable[] = {
 #if defined(USE_THROTTLE_BOOST)
     { "throttle_boost",             VAR_UINT8 | PROFILE_VALUE,  .config.minmax = { 0, 100 }, PG_PID_PROFILE, offsetof(pidProfile_t, throttle_boost) },
     { "throttle_boost_cutoff",      VAR_UINT8 | PROFILE_VALUE,  .config.minmax = { 5, 50 }, PG_PID_PROFILE, offsetof(pidProfile_t, throttle_boost_cutoff) },
+    { "anti_gravity_new",           VAR_UINT8 | PROFILE_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_OFF_ON }, PG_PID_PROFILE, offsetof(pidProfile_t, anti_gravity_new) },
```
I'm not too sure if this is a temporary variable or something similar. Use a "new" variable with ON/OFF to me is not a good idea. Is better to have a list of options for anti gravity (for example THRESHOLD, FILTER) this is better if they appear more options in a future and you can remember how it works when you see it.

ctzsnooze commented:
As per @McGiverGim's suggestion, the user  can now choose, via CLI, to have antigravity off, step (old threshold based size of step method), or smooth (new proportional to throttle rate of change method).
I've also squashed and rebased to 3.4 final.
Hopefully haven't broken anything  :-) comments welcome.

mikeller commented on src/main/fc/fc_core.c:
```diff
@@ -853,6 +853,9 @@ bool processRx(timeUs_t currentTimeUs)
     }
 #endif
+
+    pidSetAntiGravityState(IS_RC_MODE_ACTIVE(BOXANTIGRAVITY) || feature(FEATURE_ANTI_GRAVITY));
```
Why was this moved here? Doing this in `rc_modes.c` / `fc_rc.c` like it was seems to be more appropriate.

ctzsnooze replied:
I'd have to ask @etracer65 :-)

mikeller replied:
@ctzsnooze, @etracer65: I am still waiting for an answer to this.

ctzsnooze replied:
hopefully eTracer can explain why.  otherwise I'm happy to implement whatever suggestion solves the problem, it's over my head I'm afraid.

etracer65 replied:
Sorry, for some reason I wasn't receiving notifications on replies to this PR.

It's been a while so having to figure it out again, but if I recall the rationale was that the old anti-gravity was driven by the logic in fc_rc.c checking the feature/box state along with throttle movement and pushing a multiplier into the pid process using `pidSetItermAccelerator()`.  The new logic is completely in pid.c but also needs to know the active state (feature or box mode) but doesn't care about the multipler.  So the logic was normalized so that in fc_core.c (running at rx update rate) we push a flag into the PID controller that's used for both old and new modes.  This way the PID controller running at pid loop speed doesn't need to check the feature/box mode.  Also improves performance of the old mode since it was also previously checking the feature/box mode at pid loop speed in fc_rc.c `processRcCommand()`.

So in the big picture we're "pushing" the state into the PID controller but at RX update speed rather than "pulling" the state at PID loop speed.

ctzsnooze replied:
Thanks!  You're awesome, that was way over my head :-)

mikeller commented on src/main/flight/pid.c:
```diff
@@ -180,9 +189,9 @@ void pidSetItermAccelerator(float newItermAccelerator)
+bool pidOsdAntiGravityActive(void)
 {
-    return itermAccelerator;
+    return (itermAccelerator > antiGravityOSDCutoff);
```
Camel case is `antiGravityOsdCutoff`.

ctzsnooze replied:
Thanks, fixed.

jirif commented on src/main/flight/pid.h:
```diff
@@ -76,6 +76,12 @@ typedef struct pid8_s {
 } pid8_t;
 
+typedef enum {
+    ANTI_GRAVITY_OFF = 0,
```
To turn off antigravity we have a feature or switch box, why have third place where it can be turned on of off, its just mode so leave the option to handle only STEP or SMOOTH mode

ctzsnooze replied:
OK that makes sense

jirif commented on src/main/flight/pid.h:
```diff
@@ -76,6 +76,12 @@ typedef struct pid8_s {
+typedef enum {
+    ANTI_GRAVITY_OFF = 0,
+    ANTI_GRAVITY_STEP,
+    ANTI_GRAVITY_SMOOTH
```
Imho better name for SMOOTH mode should be FILTER mode

ctzsnooze replied:
I don't mind the *internal* name being ANTI_GRAVITY_FILTER. However, the outcome is a continuously variable increase in I in proportion to an increasing rate of change of throttle.  This provides a smoother range of effect that doesn't happen with the all-or-none threshold based method.  I'd prefer the user name in the CLI to reflect what the user experience will be, not the technical method used to achieve that result.

jirif replied:
Ok no problem i think entire STEP mode will be already removed soon to save flash space because yor SMOOTH mode working well and its the way to go ;)

jirif commented on src/main/flight/pid.c:
```diff
@@ -778,6 +811,13 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
 
     // Dynamic i component,
     // gradually scale back integration when above windup point
```
This comment is for dynCi variable so move it down before `const float dynCi = ...`

ctzsnooze replied:
Yes, missed that - thanks!

ctzsnooze commented:
@mikeller - wondering about what changes are needed to move forward?
